python_ta/transforms/type_inference_visitor.py

```diff
@@ -321,48 +325,46 @@ def visit_return(self, node):
         node.type_constraints = TypeInfo(NoType)
 
     def visit_functiondef(self, node):
-        arg_types = [self.type_constraints.lookup_concrete(node.type_environment.lookup_in_env(arg))
-                     for arg in node.argnames()]
+        arg_types = [self.lookup_type(node, arg) for arg in node.argnames()]
         if any(annotation is not None for annotation in node.args.annotations):
             func_type = parse_annotations(node)
             for arg_type, annotation in zip(arg_types, func_type.__args__[:-1]):
                 self.type_constraints.unify(arg_type, annotation)
-            self.type_constraints.unify(self.lookup_type(node, node.name), func_type)
+            self.type_constraints.unify(self.lookup_type(node.parent, node.name), func_type)
         else:
             # Check whether this is a method in a class
             if isinstance(node.parent, astroid.ClassDef) and isinstance(arg_types[0], TypeVar):
-                self.type_constraints.unify(arg_types[0], _ForwardRef(node.parent.name))
+                self.type_constraints.unify(arg_types[0], _ForwardRef(node.parent.name), node)
 
             # check if return nodes exist; there is a return statement in function body.
             if len(list(node.nodes_of_class(astroid.Return))) == 0:
                 func_type = Callable[arg_types, None]
             else:
+                polymorphic_tvars = [arg for arg in arg_types if isinstance(arg, TypeVar)]
```

Comment:
This part should really be done for both branches of the `if`. The `create_Callable` is a good change to make, and should supersede the code in the `if` branch.